### DISCUSSION TITLE
JSON Parser only outputs "null" values when using variables in strings

### DISCUSSION THREAD

**Discussion by add-n2x** [savonet/liquidsoap#2484] [User Support]
**Describe the bug**
The new Liquidsoap 2.1 JSON Parser doesn't work when including variables in strings. 

**To Reproduce**

## Scenario 1: Doesn't work

```ini

let json.parse data = '{"show_id": "-1", "show_name": "Random Music"}'
print("Set current show to '#{data.show_name}' (ID: #{data.show_id})")

# > Set current show to 'null' (ID: null)
```

## Scenario 2: Works

```ini
let json.parse data = '{"show_id": "-1", "show_name": "Random Music"}'
print("Set current show to '" ^ data.show_name ^ "' (ID: " ^ data.show_id ^ ")")

# > Set current show to 'Random Music' (ID: -1)
```

## Scenario 3: Works, when properties have been accessed before using a var string

```ini
let json.parse data = '{"show_id": "-1", "show_name": "Random Music"}'
print("Set current show to '" ^ data.show_name ^ "' (ID: " ^ data.show_id ^ ")")
print("Set current show to '#{data.show_name}' (ID: #{data.show_id})")

# > Set current show to 'Random Music' (ID: -1)
# > Set current show to 'Random Music' (ID: -1)
```

**Expected behavior**

Scenario 1.) would print `Set current show to 'Random Music' (ID: -1)`

**Version details**
 - OS: Ubuntu 21.10
 - Liquidsoap 2.1.x - rolling release

**Install method**
- Rolling release, Debian package

**Comment by toots** [comment]
Thanks for trying the new parser! 

This behavior is unfortunately expected and explained in the documentation: https://www.liquidsoap.info/doc-dev/json.html

The reason is that the string interpolation `#{..}` does not force its content to be a `string` because all values can be converted to `string` in liquidsoap (yey polymorphism madness).

The recommended way to parse JSON should always be with an associated type annotation:
```ruby
let json.parse data : { show_id: string, show_name: string } = '{"show_id": "-1", "show_name": "Random Music"}'
```
Or, even:
```ruby
let json.parse { show_id, show_name } : { show_id: string, show_name: string } = '{"show_id": "-1", "show_name": "Random Music"}'
```

I'm gonna add a warning when parsing w/o type annotation so that is clear to the user. Thanks for reporting!

**Reply by add-n2x** [reply]
Got ya! Thanks for the quick reply.

**Reply by add-n2x** [reply]
Still, when trying your example

`let json.parse data : { show_id: string, show_name: string } = '{"show_id": "-1", "show_name": "Random Music"}'`

I get:

```
At test_parse_json.liq, line 32, char 20:
let json.parse data : { show_id: string, show_name: string } = '{"show_id": "-1", "show_name": "Random Music"}'

Error 2: Parse error
```

highlighting the first ":" colon as source of error.

**Reply by toots** [reply]
You need parenthesis:
```ruby
let json.parse (data : {
  show_id: string,
  show_name: string
}) = '{"show_id": "-1", "show_name": "Random Music"}'
```

**Reply by toots** [reply]
BTW we have a rudimentary JS runtime for testing these things now! Here: https://www.liquidsoap.info/try/

**Reply by add-n2x** [reply]
With parenthesis it does indeed work. Thanks.

> BTW we have a rudimentary JS runtime for testing these things now! Here: https://www.liquidsoap.info/try/

Oh wow, that's really handy. Great to see you come up with additional tooling. :-)
